### DISCUSSION TITLE
Working as standalone without HA

### DISCUSSION THREAD

**Discussion by 0rsa** [Tasshack/dreame-vacuum#587] [General]
Hello,

First, thanks for the library and the hard work done (This is the only library that I know of that allows you to use a Dreame account).
Since I'm not using Home Assistant, I'm trying to use the dreame module separately with a custom script.

I'm trying to instantiate a new DreameVacuumDevice using the latest beta released on yesterday.

device = DreameVacuumDevice(params)

then I call the update method to init the connection.

device.update()

It connects successfully to the cloud and to the device but it fails at these lines in the update() method

> if not self.device_connected:
>             raise DeviceUpdateFailedException("Device cannot be reached") from None

Even it the connection is a success in the logs :

> DEBUG:dreame.device:Device update: 5
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): cn.iot.dreame.tech:13267
> DEBUG:urllib3.connectionpool:https://cn.iot.dreame.tech:13267 "POST /dreame-auth/oauth/token HTTP/1.1" 200 None
> DEBUG:urllib3.connectionpool:https://cn.iot.dreame.tech:13267 "POST /dreame-user-iot/iotuserbind/device/listV2 HTTP/1.1" 200 None
> INFO:dreame.device:Connecting to device
> DEBUG:urllib3.connectionpool:https://cn.iot.dreame.tech:13267 "POST /dreame-user-iot/iotuserbind/device/info HTTP/1.1" 200 None
> DEBUG:urllib3.connectionpool:https://cn.iot.dreame.tech:13267 "POST /dreame-user-iot/iotstatus/devOTCInfo HTTP/1.1" 200 None
> INFO:dreame.protocol:Connecting to the device client
> INFO:dreame.device:Connected to device: dreame.vacuum.r2401 4.3.9_1187

I'm not able to find where the "device_connected" attribute is set to true in the process.

Can you help me to go in the right way? I think and I hope it can be useful for people who don't use HA as domotic core.

Thank you! 😄

**Comment by Tasshack** [comment]
https://github.com/Tasshack/dreame-vacuum/blob/dev/custom_components/dreame_vacuum/dreame/device.py#L5392
https://github.com/Tasshack/dreame-vacuum/blob/dev/custom_components/dreame_vacuum/dreame/protocol.py#L151

**Comment by 0rsa** [comment]
Thank you, I finally found why the _client_connected is still false with your help.

In the connect method of protocol.py, an exception is thrown:
> An exception occurred: Unsupported callback API version: version 2.0 added a callback_api_version, see docs/migrations.rst for details

Googling this one, the error was caused by the paho-mqtt version which is > 2.0.0 in my case.

Reinstalling an older version solves the issue :
pip install "paho-mqtt<2.0.0"

I keep going to play with the module without HA, thank you!

**Comment by Tasshack** [comment]
@0rsa I am going to refactor whole connected/disconnected state handing because currently it is not working correctly and mqtt connection only guarantees the cloud connection but not connection to the device since it's just a cloud relay for getting the updates from device.
I don't really recommend trying to port the complete module because althought it looks like an independed, lots of backend logic handled by Home Assistant entities like in the `entity.py`. Also current `device.py` heavily relies on my external custom built development tools for extracting capability information from individual app plugins. Almost all of the logic has been reversed engineered from the app but it would be much easier if Home Assistant had a state like `disabled`.

What I recommend to you is using the `protocol.py` for accessing the Dreame cloud and using the `map.py` for accessing the map data/rendering and develop your own device logic with your requirements.
